storage/local/storage.go

```diff
@@ -888,40 +963,42 @@ func (s *MemorySeriesStorage) seriesForRange(
 }
 
 func (s *MemorySeriesStorage) preloadChunksForRange(
-	fp model.Fingerprint,
+	fpsPair fingerprintSeriesPair,
 	from model.Time, through model.Time,
 ) SeriesIterator {
-	s.fpLocker.Lock(fp)
-	defer s.fpLocker.Unlock(fp)
-
-	series := s.seriesForRange(fp, from, through)
+	fp, series := fpsPair.fp, fpsPair.series
 	if series == nil {
 		return nopIter
 	}
+
+	s.fpLocker.Lock(fp)
```

Comment:
As discussed in the other PR, the performance gain of not going with `defer` is probably negligible in this context, and will go away with Go1.8 anyway. Better use the cleaner code here (i.e. unlock with defer as before).